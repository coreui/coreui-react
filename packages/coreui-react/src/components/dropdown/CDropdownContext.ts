import { createContext, RefObject } from 'react'
import { Alignments } from './types'

export interface CDropdownContextProps {
  alignment?: Alignments
  container?: DocumentFragment | Element | (() => DocumentFragment | Element | null) | null
  dark?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dropdownToggleRef: RefObject<any | null>
  dropdownMenuRef: RefObject<HTMLDivElement | HTMLUListElement | null>
  setVisible: React.Dispatch<React.SetStateAction<boolean | undefined>>
  popper?: boolean
  portal?: boolean
  variant?: 'btn-group' | 'dropdown' | 'input-group' | 'nav-item'
  visible?: boolean
}

export const CDropdownContext = createContext({} as CDropdownContextProps)
