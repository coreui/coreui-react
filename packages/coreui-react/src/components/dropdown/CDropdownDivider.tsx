import React, { forwardRef, HTMLAttributes } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

export interface CDropdownDividerProps extends HTMLAttributes<HTMLHRElement> {
  /**
   * A string of all className you want applied to the component.
   */
  className?: string
}

export const CDropdownDivider = forwardRef<HTMLHRElement, CDropdownDividerProps>(
  ({ className, ...rest }, ref) => {
    return <hr className={classNames('dropdown-divider', className)} {...rest} ref={ref} />
  }
)

CDropdownDivider.propTypes = {
  className: PropTypes.string,
}

CDropdownDivider.displayName = 'CDropdownDivider'
