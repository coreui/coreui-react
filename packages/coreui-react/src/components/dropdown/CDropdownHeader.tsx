import React, { ElementType, forwardRef, HTMLAttributes } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { PolymorphicRefForwardingComponent } from '../../helpers'

export interface CDropdownHeaderProps extends HTMLAttributes<HTMLHeadingElement> {
  /**
   * Component used for the root node. Either a string to use a HTML element or a component.
   */
  as?: ElementType
  /**
   * A string of all className you want applied to the component.
   */
  className?: string
}

export const CDropdownHeader: PolymorphicRefForwardingComponent<'h6', CDropdownHeaderProps> =
  forwardRef<HTMLHeadingElement, CDropdownHeaderProps>(
    ({ children, as: Component = 'h6', className, ...rest }, ref) => {
      return (
        <Component className={classNames('dropdown-header', className)} {...rest} ref={ref}>
          {children}
        </Component>
      )
    }
  )

CDropdownHeader.propTypes = {
  as: PropTypes.elementType,
  children: PropTypes.node,
  className: PropTypes.string,
}

CDropdownHeader.displayName = 'CDropdownHeader'
