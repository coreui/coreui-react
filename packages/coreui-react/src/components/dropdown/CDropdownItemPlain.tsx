import React, { ElementType, forwardRef, HTMLAttributes } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { PolymorphicRefForwardingComponent } from '../../helpers'

export interface CDropdownItemPlainProps extends HTMLAttributes<HTMLSpanElement> {
  /**
   * Component used for the root node. Either a string to use a HTML element or a component.
   */
  as?: ElementType
  /**
   * A string of all className you want applied to the component.
   */
  className?: string
}

export const CDropdownItemPlain: PolymorphicRefForwardingComponent<
  'span',
  CDropdownItemPlainProps
> = forwardRef<HTMLSpanElement, CDropdownItemPlainProps>(
  ({ children, as: Component = 'span', className, ...rest }, ref) => {
    return (
      <Component className={classNames('dropdown-item-text', className)} {...rest} ref={ref}>
        {children}
      </Component>
    )
  }
)

CDropdownItemPlain.propTypes = {
  as: PropTypes.elementType,
  children: PropTypes.node,
  className: PropTypes.string,
}

CDropdownItemPlain.displayName = 'CDropdownItemPlain'
