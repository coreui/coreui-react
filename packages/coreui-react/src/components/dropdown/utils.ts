import type { Placement } from '@popperjs/core'
import type { Placements } from '../../types'
import type { Alignments, Breakpoints } from './types'

export const getAlignmentClassNames = (alignment: Alignments) => {
  const classNames: string[] = []
  if (typeof alignment === 'object') {
    for (const key in alignment) {
      classNames.push(
        `dropdown-menu${key === 'xs' ? '' : `-${key}`}-${alignment[key as keyof Breakpoints]}`
      )
    }
  }

  if (typeof alignment === 'string') {
    classNames.push(`dropdown-menu-${alignment}`)
  }

  return classNames
}

export const getPlacement = (
  placement: Placement,
  direction: string | undefined,
  alignment: Alignments | string | undefined,
  isRTL: boolean
): Placements => {
  let _placement = placement

  if (direction === 'dropup') {
    _placement = isRTL ? 'top-end' : 'top-start'
  }

  if (direction === 'dropup-center') {
    _placement = 'top'
  }

  if (direction === 'dropend') {
    _placement = isRTL ? 'left-start' : 'right-start'
  }

  if (direction === 'dropstart') {
    _placement = isRTL ? 'right-start' : 'left-start'
  }

  if (alignment === 'end') {
    _placement = isRTL ? 'bottom-start' : 'bottom-end'
  }

  return _placement
}
