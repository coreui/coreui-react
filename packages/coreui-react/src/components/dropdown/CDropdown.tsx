import React, {
  createContext,
  ElementType,
  forwardRef,
  HTMLAttributes,
  RefObject,
  useEffect,
  useRef,
  useState,
} from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import type { Options } from '@popperjs/core'

import { PolymorphicRefForwardingComponent } from '../../helpers'
import { useForkedRef, usePopper } from '../../hooks'
import { placementPropType } from '../../props'
import type { Placements } from '../../types'
import { getNextActiveElement, isRTL } from '../../utils'

import type { Alignments, Directions } from './types'
import { getPlacement } from './utils'

export interface CDropdownProps extends HTMLAttributes<HTMLDivElement | HTMLLIElement> {
  /**
   * Specifies the alignment of the React Dropdown Menu within this React Dropdown.
   *
   * @example
   * // Align dropdown menu to the end on large devices, otherwise start
   * <CDropdown alignment={{ lg: 'end', xs: 'start' }}>
   *   <CDropdownToggle>Toggle dropdown</CDropdownToggle>
   *   <CDropdownMenu>
   *     <CDropdownItem>Action</CDropdownItem>
   *     <CDropdownItem>Another Action</CDropdownItem>
   *   </CDropdownMenu>
   * </CDropdown>
   *
   * @type 'start' | 'end' | { xs: 'start' | 'end' } | { sm: 'start' | 'end' } | { md: 'start' | 'end' } | { lg: 'start' | 'end' } | { xl: 'start' | 'end'} | { xxl: 'start' | 'end'}
   */
  alignment?: Alignments

  /**
   * Determines the root node component (native HTML element or a custom React component) for the React Dropdown.
   */
  as?: ElementType

  /**
   * Configures automatic closing behavior for the React Dropdown:
   * - `true` - Close on clicks inside or outside of the React Dropdown Menu.
   * - `false` - Disable auto-close; manually call `hide` or `toggle` (also not closed by `Escape`).
   * - `'inside'` - Close only when clicking inside the React Dropdown Menu.
   * - `'outside'` - Close only when clicking outside the React Dropdown Menu.
   *
   * @example
   * // Close only when user clicks outside of the menu
   * <CDropdown autoClose="outside" />
   */
  autoClose?: 'inside' | 'outside' | boolean

  /**
   * Adds custom classes to the React Dropdown root element.
   */
  className?: string

  /**
   * Appends the React Dropdown Menu to a specific element. You can pass an HTML element or a function returning an element. Defaults to `document.body`.
   *
   * @example
   * // Append the menu to a custom container
   * const myContainer = document.getElementById('my-container')
   *
   * <CDropdown container={myContainer} />
   *
   * @since 4.11.0
   */
  container?: DocumentFragment | Element | (() => DocumentFragment | Element | null) | null

  /**
   * Applies a darker color scheme to the React Dropdown Menu, often used within dark navbars.
   */
  dark?: boolean

  /**
   * Specifies the direction of the React Dropdown.
   */
  direction?: 'center' | 'dropup' | 'dropup-center' | 'dropend' | 'dropstart'

  /**
   * Defines x and y offsets ([x, y]) for the React Dropdown Menu relative to its target.
   *
   * @example
   * // Offset the menu 10px in X and 5px in Y direction
   * <CDropdown offset={[10, 5]}>
   *   ...
   * </CDropdown>
   */
  offset?: [number, number]

  /**
   * Callback fired right before the React Dropdown becomes hidden.
   *
   * @since 4.9.0
   */
  onHide?: () => void

  /**
   * Callback fired immediately after the React Dropdown is displayed.
   */
  onShow?: () => void

  /**
   * Determines the placement of the React Dropdown Menu after Popper.js modifiers.
   *
   * @type 'auto' | 'auto-start' | 'auto-end' | 'top-end' | 'top' | 'top-start' | 'bottom-end' | 'bottom' | 'bottom-start' | 'right-start' | 'right' | 'right-end' | 'left-start' | 'left' | 'left-end'
   */
  placement?: Placements

  /**
   * Enables or disables dynamic positioning via Popper.js for the React Dropdown Menu.
   */
  popper?: boolean

  /**
   * Provides a custom Popper.js configuration or a function that returns a modified Popper.js configuration for advanced positioning of the React Dropdown Menu. [Read more](https://popper.js.org/docs/v2/constructors/#options)
   *
   * @example
   * // Providing a custom popper config
   * <CDropdown
   *   popperConfig={{
   *     modifiers: [
   *       {
   *         name: 'flip',
   *         options: { fallbackPlacements: ['bottom', 'top'] },
   *       },
   *     ],
   *   }}
   * >...</CDropdown>
   *
   * @since 5.5.0
   */
  popperConfig?: Partial<Options> | ((defaultPopperConfig: Partial<Options>) => Partial<Options>)

  /**
   * Renders the React Dropdown Menu using a React Portal, allowing it to escape the DOM hierarchy for improved positioning.
   *
   * @since 4.8.0
   */
  portal?: boolean

  /**
   * Defines the visual variant of the React Dropdown
   */
  variant?: 'btn-group' | 'dropdown' | 'input-group' | 'nav-item'

  /**
   * Controls the visibility of the React Dropdown Menu:
   * - `true` - Visible
   * - `false` - Hidden
   *
   * @example
   * // Programmatically manage the dropdown visibility
   * const [visible, setVisible] = useState(false)
   *
   * <CDropdown visible={visible}>
   *   ...
   * </CDropdown>
   *
   */
  visible?: boolean
}

interface ContextProps extends CDropdownProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dropdownToggleRef: RefObject<any | undefined>
  dropdownMenuRef: RefObject<HTMLDivElement | HTMLUListElement | undefined>
  setVisible: React.Dispatch<React.SetStateAction<boolean | undefined>>
  portal: boolean
}

export const CDropdownContext = createContext({} as ContextProps)

export const CDropdown: PolymorphicRefForwardingComponent<'div', CDropdownProps> = forwardRef<
  HTMLDivElement | HTMLLIElement,
  CDropdownProps
>(
  (
    {
      children,
      alignment,
      as = 'div',
      autoClose = true,
      className,
      container,
      dark,
      direction,
      offset = [0, 2],
      onHide,
      onShow,
      placement = 'bottom-start',
      popper = true,
      popperConfig,
      portal = false,
      variant = 'btn-group',
      visible = false,
      ...rest
    },
    ref
  ) => {
    const dropdownRef = useRef<HTMLDivElement>(null)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const dropdownToggleRef = useRef<any>(null)
    const dropdownMenuRef = useRef<HTMLDivElement | HTMLUListElement>(null)
    const forkedRef = useForkedRef(ref, dropdownRef)
    const [_visible, setVisible] = useState(visible)
    const { initPopper, destroyPopper } = usePopper()

    const Component = variant === 'nav-item' ? 'li' : as

    // Disable popper if responsive aligment is set.
    if (typeof alignment === 'object') {
      popper = false
    }

    const contextValues = {
      alignment,
      container,
      dark,
      dropdownToggleRef,
      dropdownMenuRef,
      popper,
      portal,
      variant,
      visible: _visible,
      setVisible,
    }

    const defaultPopperConfig = {
      modifiers: [
        {
          name: 'offset',
          options: {
            offset: offset,
          },
        },
      ],
      placement: getPlacement(placement, direction, alignment, isRTL(dropdownMenuRef.current)),
    }

    const computedPopperConfig: Partial<Options> = {
      ...defaultPopperConfig,
      ...(typeof popperConfig === 'function' ? popperConfig(defaultPopperConfig) : popperConfig),
    }

    useEffect(() => {
      setVisible(visible)
    }, [visible])

    useEffect(() => {
      if (_visible && dropdownToggleRef.current && dropdownMenuRef.current) {
        dropdownToggleRef.current.focus()
        popper &&
          initPopper(dropdownToggleRef.current, dropdownMenuRef.current, computedPopperConfig)
        window.addEventListener('mouseup', handleMouseUp)
        window.addEventListener('keyup', handleKeyup)
        dropdownToggleRef.current.addEventListener('keydown', handleKeydown)
        dropdownMenuRef.current.addEventListener('keydown', handleKeydown)
        onShow && onShow()
      }

      return () => {
        popper && destroyPopper()
        window.removeEventListener('mouseup', handleMouseUp)
        window.removeEventListener('keyup', handleKeyup)
        dropdownToggleRef.current &&
          dropdownToggleRef.current.removeEventListener('keydown', handleKeydown)
        dropdownMenuRef.current &&
          dropdownMenuRef.current.removeEventListener('keydown', handleKeydown)
        onHide && onHide()
      }
    }, [_visible])

    const handleKeydown = (event: KeyboardEvent) => {
      if (
        _visible &&
        dropdownMenuRef.current &&
        (event.key === 'ArrowDown' || event.key === 'ArrowUp')
      ) {
        event.preventDefault()
        const target = event.target as HTMLElement
        const items: HTMLElement[] = Array.from(
          dropdownMenuRef.current.querySelectorAll('.dropdown-item:not(.disabled):not(:disabled)')
        )
        getNextActiveElement(items, target, event.key === 'ArrowDown', true).focus()
      }
    }

    const handleKeyup = (event: KeyboardEvent) => {
      if (autoClose === false) {
        return
      }

      if (event.key === 'Escape') {
        setVisible(false)
      }
    }

    const handleMouseUp = (event: Event) => {
      if (!dropdownToggleRef.current || !dropdownMenuRef.current) {
        return
      }

      if (dropdownToggleRef.current.contains(event.target as HTMLElement)) {
        return
      }

      if (
        autoClose === true ||
        (autoClose === 'inside' && dropdownMenuRef.current.contains(event.target as HTMLElement)) ||
        (autoClose === 'outside' && !dropdownMenuRef.current.contains(event.target as HTMLElement))
      ) {
        setTimeout(() => setVisible(false), 1)
        return
      }
    }

    return (
      <CDropdownContext.Provider value={contextValues}>
        {variant === 'input-group' ? (
          <>{children}</>
        ) : (
          <Component
            className={classNames(
              variant === 'nav-item' ? 'nav-item dropdown' : variant,
              {
                'dropdown-center': direction === 'center',
                'dropup dropup-center': direction === 'dropup-center',
                [`${direction}`]:
                  direction && direction !== 'center' && direction !== 'dropup-center',
              },
              className
            )}
            {...rest}
            ref={forkedRef}
          >
            {children}
          </Component>
        )}
      </CDropdownContext.Provider>
    )
  }
)

const alignmentDirection = PropTypes.oneOf<Directions>(['start', 'end'])

CDropdown.propTypes = {
  alignment: PropTypes.oneOfType([
    alignmentDirection,
    PropTypes.shape({ xs: alignmentDirection.isRequired }),
    PropTypes.shape({ sm: alignmentDirection.isRequired }),
    PropTypes.shape({ md: alignmentDirection.isRequired }),
    PropTypes.shape({ lg: alignmentDirection.isRequired }),
    PropTypes.shape({ xl: alignmentDirection.isRequired }),
    PropTypes.shape({ xxl: alignmentDirection.isRequired }),
  ]),
  as: PropTypes.elementType,
  autoClose: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.oneOf<'inside' | 'outside'>(['inside', 'outside']),
  ]),
  children: PropTypes.node,
  className: PropTypes.string,
  dark: PropTypes.bool,
  direction: PropTypes.oneOf(['center', 'dropup', 'dropup-center', 'dropend', 'dropstart']),
  offset: PropTypes.any, // TODO: find good proptype
  onHide: PropTypes.func,
  onShow: PropTypes.func,
  placement: placementPropType,
  popper: PropTypes.bool,
  popperConfig: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  portal: PropTypes.bool,
  variant: PropTypes.oneOf(['btn-group', 'dropdown', 'input-group', 'nav-item']),
  visible: PropTypes.bool,
}

CDropdown.displayName = 'CDropdown'
