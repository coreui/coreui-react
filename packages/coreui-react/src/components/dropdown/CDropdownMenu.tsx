import React, { ElementType, forwardRef, HTMLAttributes, useContext } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { CDropdownContext } from './CDropdown'
import { CConditionalPortal } from '../conditional-portal'

import { PolymorphicRefForwardingComponent } from '../../helpers'
import { useForkedRef } from '../../hooks'

import { getAlignmentClassNames } from './utils'

export interface CDropdownMenuProps extends HTMLAttributes<HTMLDivElement | HTMLUListElement> {
  /**
   * Component used for the root node. Either a string to use a HTML element or a component.
   */
  as?: ElementType
  /**
   * A string of all className you want applied to the base component.
   */
  className?: string
}

export const CDropdownMenu: PolymorphicRefForwardingComponent<'ul', CDropdownMenuProps> =
  forwardRef<HTMLDivElement | HTMLUListElement, CDropdownMenuProps>(
    ({ children, as: Component = 'ul', className, ...rest }, ref) => {
      const { alignment, container, dark, dropdownMenuRef, popper, portal, visible } =
        useContext(CDropdownContext)

      const forkedRef = useForkedRef(ref, dropdownMenuRef)

      return (
        <CConditionalPortal container={container} portal={portal ?? false}>
          <Component
            className={classNames(
              'dropdown-menu',
              {
                show: visible,
              },
              alignment && getAlignmentClassNames(alignment),
              className,
            )}
            ref={forkedRef}
            role="menu"
            {...(!popper && { 'data-coreui-popper': 'static' })}
            {...(dark && { 'data-coreui-theme': 'dark' })}
            {...rest}
          >
            {Component === 'ul'
              ? React.Children.map(children, (child, index) => {
                  if (React.isValidElement(child)) {
                    return <li key={index}>{React.cloneElement(child)}</li>
                  }
                  return
                })
              : children}
          </Component>
        </CConditionalPortal>
      )
    },
  )

CDropdownMenu.propTypes = {
  as: PropTypes.elementType,
  children: PropTypes.node,
  className: PropTypes.string,
}

CDropdownMenu.displayName = 'CDropdownMenu'
