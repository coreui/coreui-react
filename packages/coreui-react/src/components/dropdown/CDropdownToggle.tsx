import React, { FC, useContext } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { CButton, CButtonProps } from '../button/CButton'

import { CDropdownContext } from './CDropdown'

import { triggerPropType } from '../../props'
import type { Triggers } from '../../types'

export interface CDropdownToggleProps extends Omit<CButtonProps, 'type'> {
  /**
   * Enables pseudo element caret on toggler.
   */
  caret?: boolean
  /**
   * Create a custom toggler which accepts any content.
   */
  custom?: boolean
  /**
   * If a dropdown `variant` is set to `nav-item` then render the toggler as a link instead of a button.
   *
   * @since 5.0.0
   */
  navLink?: boolean
  /**
   * Similarly, create split button dropdowns with virtually the same markup as single button dropdowns, but with the addition of `.dropdown-toggle-split` className for proper spacing around the dropdown caret.
   */
  split?: boolean
  /**
   * Sets which event handlers youâ€™d like provided to your toggle prop. You can specify one trigger or an array of them.
   *
   * @type 'hover' | 'focus' | 'click'
   */
  trigger?: Triggers | Triggers[]
}

export const CDropdownToggle: FC<CDropdownToggleProps> = ({
  children,
  caret = true,
  custom,
  className,
  navLink = true,
  split,
  trigger = 'click',
  ...rest
}) => {
  const { dropdownToggleRef, variant, visible, setVisible } = useContext(CDropdownContext)

  const triggers = {
    ...((trigger === 'click' || trigger.includes('click')) && {
      onClick: (event: React.MouseEvent<HTMLElement>) => {
        event.preventDefault()
        setVisible(!visible)
      },
    }),
    ...((trigger === 'focus' || trigger.includes('focus')) && {
      onFocus: () => setVisible(true),
      onBlur: () => setVisible(false),
    }),
  }

  const togglerProps = {
    className: classNames(
      {
        'nav-link': variant === 'nav-item' && navLink,
        'dropdown-toggle': caret,
        'dropdown-toggle-split': split,
        show: visible,
      },
      className,
    ),
    'aria-expanded': visible,
    ...(!rest.disabled && { ...triggers }),
  }

  const Toggler = () => {
    if (custom && React.isValidElement(children)) {
      return (
        <>
          {React.cloneElement(children as React.ReactElement<any>, {
            'aria-expanded': visible,
            ...(!rest.disabled && { ...triggers }),
            ref: dropdownToggleRef,
          })}
        </>
      )
    }

    if (variant === 'nav-item' && navLink) {
      return (
        <a href="#" {...togglerProps} role="button" {...rest} ref={dropdownToggleRef}>
          {children}
        </a>
      )
    }

    return (
      <CButton {...togglerProps} tabIndex={0} {...rest} ref={dropdownToggleRef}>
        {children}
        {split && <span className="visually-hidden">Toggle Dropdown</span>}
      </CButton>
    )
  }

  return <Toggler />
}

CDropdownToggle.propTypes = {
  caret: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  custom: PropTypes.bool,
  split: PropTypes.bool,
  trigger: triggerPropType,
}

CDropdownToggle.displayName = 'CDropdownToggle'
