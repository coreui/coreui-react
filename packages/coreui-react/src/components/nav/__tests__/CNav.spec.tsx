import * as React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { CDropdown, CDropdownItem, CDropdownMenu, CDropdownToggle } from '../../dropdown'
import { CNav, CNavItem, CNavLink } from '../index'

test('loads and displays CNav component', async () => {
  const { container } = render(<CNav>Test</CNav>)
  expect(container).toMatchSnapshot()
})

test('CNav customize', async () => {
  const { container } = render(
    <CNav className="bazinga" as="h3" layout="justified" variant="pills">
      Test
    </CNav>
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('nav')
  expect(container.firstChild).toHaveClass('nav-justified')
  expect(container.firstChild).toHaveClass('nav-pills')
  expect(container.firstChild).toHaveClass('bazinga')
})

test('CNav example', async () => {
  const { container } = render(
    <CNav>
      <CNavItem>
        <CNavLink href="#" active>
          Active
        </CNavLink>
      </CNavItem>
      <CNavItem>
        <CNavLink href="#">Link</CNavLink>
      </CNavItem>
      <CDropdown variant="nav-item">
        <CDropdownToggle>A</CDropdownToggle>
        <CDropdownMenu>
          <CDropdownItem href="#">B</CDropdownItem>
          <CDropdownItem href="#">C</CDropdownItem>
          <CDropdownItem href="#">D</CDropdownItem>
        </CDropdownMenu>
      </CDropdown>
      <CNavItem>
        <CNavLink href="#">Link</CNavLink>
      </CNavItem>
      <CNavItem>
        <CNavLink href="#" disabled>
          Disabled
        </CNavLink>
      </CNavItem>
    </CNav>
  )
  expect(container).toMatchSnapshot()
})
