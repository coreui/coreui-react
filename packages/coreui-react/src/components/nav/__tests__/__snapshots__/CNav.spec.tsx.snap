// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CNav customize 1`] = `
<div>
  <h3
    class="nav nav-justified nav-pills bazinga"
    role="navigation"
  >
    Test
  </h3>
</div>
`;

exports[`CNav example 1`] = `
<div>
  <ul
    class="nav"
    role="navigation"
  >
    <li
      class="nav-item"
    >
      <a
        aria-current="page"
        class="nav-link active"
        href="#"
      >
        Active
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="#"
      >
        Link
      </a>
    </li>
    <li
      class="nav-item dropdown"
    >
      <a
        aria-expanded="false"
        class="nav-link dropdown-toggle"
        href="#"
        role="button"
      >
        A
      </a>
      <ul
        class="dropdown-menu"
        role="menu"
      >
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            B
          </a>
        </li>
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            C
          </a>
        </li>
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            D
          </a>
        </li>
      </ul>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="#"
      >
        Link
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        aria-disabled="true"
        class="nav-link disabled"
        disabled=""
        href="#"
        tabindex="-1"
      >
        Disabled
      </a>
    </li>
  </ul>
</div>
`;

exports[`loads and displays CNav component 1`] = `
<div>
  <ul
    class="nav"
    role="navigation"
  >
    Test
  </ul>
</div>
`;
